controllers/controllers/services/instances/upsi/controller.go

```diff
@@ -133,11 +133,29 @@ func (r *Reconciler) ReconcileResource(ctx context.Context, cfServiceInstance *k
 	}
 
 	if err = r.validateCredentials(credentialsSecret); err != nil {
+		cfServiceInstance.Status.LastOperation = korifiv1alpha1.LastOperation{
+			Type:  "create",
+			State: "failed",
+		}
 		return ctrl.Result{}, k8s.NewNotReadyError().WithCause(err).WithReason("SecretInvalid")
 	}
 
 	log.V(1).Info("credentials secret", "name", credentialsSecret.Name, "version", credentialsSecret.ResourceVersion)
 	cfServiceInstance.Status.Credentials = corev1.LocalObjectReference{Name: credentialsSecret.Name}
+
+	if cfServiceInstance.Status.CredentialsObservedVersion == "" {
```

Comment:
We should extract these two conditionals in a helper function that returns last operation to tidy up the main loop.